### PR TITLE
[Merged by Bors] - feat(ModelTheory/Complexity): Preservation theorems for atomic and quantifier-free formulas

### PR DESCRIPTION
**Pull request by awainverse**
Adds a bit of extra API for quantifier-free formulas
Proves some preservation theorems of atomic and quantifier-free formulas with respect to certain classes of maps, including embeddings

---

- [x] depends on: #16107

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/ModelTheory/Complexity.lean [hunk 1]
```diff
@@ -79,19 +79,31 @@ theorem IsAtomic.isQF {φ : L.BoundedFormula α n} : IsAtomic φ → IsQF φ :=
...
-theorem IsQF.relabel {m : ℕ} {φ : L.BoundedFormula α m} (h : φ.IsQF) (f : α → β ⊕ (Fin n)) :
+theorem top : IsQF (⊤ : L.BoundedFormula α n) := isQF_bot.not
+
+theorem sup {φ ψ : L.BoundedFormula α n} (hφ : IsQF φ) (hψ : IsQF ψ) : IsQF (φ ⊔ ψ) :=
+  hφ.not.imp hψ
+
+theorem inf {φ ψ : L.BoundedFormula α n} (hφ : IsQF φ) (hψ : IsQF ψ) : IsQF (φ ⊓ ψ) :=
+  (hφ.imp hψ.not).not
+
+theorem relabel {m : ℕ} {φ : L.BoundedFormula α m} (h : φ.IsQF) (f : α → β ⊕ (Fin n)) :
```
Comment (YaelDillies) [inline Mathlib/ModelTheory/Complexity.lean:95]:
```suggestion
protected theorem relabel {m : ℕ} {φ : L.BoundedFormula α m} (h : φ.IsQF) (f : α → β ⊕ (Fin n)) :
```

File: Mathlib/ModelTheory/Complexity.lean [hunk 2]
```diff
@@ -79,19 +79,31 @@ theorem IsAtomic.isQF {φ : L.BoundedFormula α n} : IsAtomic φ → IsQF φ :=
...
     (φ.relabel f).IsQF :=
   IsQF.recOn h isQF_bot (fun h => (h.relabel f).isQF) fun _ _ h1 h2 => h1.imp h2
 
-theorem IsQF.liftAt {k m : ℕ} (h : IsQF φ) : (φ.liftAt k m).IsQF :=
+theorem liftAt {k m : ℕ} (h : IsQF φ) : (φ.liftAt k m).IsQF :=
```
Comment (YaelDillies) [inline Mathlib/ModelTheory/Complexity.lean:99]:
```suggestion
protected theorem liftAt {k m : ℕ} (h : IsQF φ) : (φ.liftAt k m).IsQF :=
```

File: Mathlib/ModelTheory/Complexity.lean [hunk 3]
```diff
@@ -79,19 +79,31 @@ theorem IsAtomic.isQF {φ : L.BoundedFormula α n} : IsAtomic φ → IsQF φ :=
...
   IsQF.recOn h isQF_bot (fun ih => ih.liftAt.isQF) fun _ _ ih1 ih2 => ih1.imp ih2
 
-theorem IsQF.castLE {h : l ≤ n} (hφ : IsQF φ) : (φ.castLE h).IsQF :=
+theorem castLE {h : l ≤ n} (hφ : IsQF φ) : (φ.castLE h).IsQF :=
```
Comment (YaelDillies) [inline Mathlib/ModelTheory/Complexity.lean:102]:
```suggestion
protected theorem castLE {h : l ≤ n} (hφ : IsQF φ) : (φ.castLE h).IsQF :=
```

**Review by YaelDillies — COMMENTED** [review]
Can you merge master to reduce the diff?

**Review by YaelDillies — APPROVED** [review]
maintainer merge

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**